Makefile

```diff
@@ -1,60 +1,67 @@
-VENV = venv
-PYTHON = $(VENV)/bin/python3
-PIP = $(VENV)/bin/pip
+PYTHON = python3
 
+VENV      = venv
+ACTIVATE := source $(VENV)/bin/activate
 
-$(VENV)/bin/activate: requirements.txt
-	python3 -m venv $(VENV)
-	$(PIP) install --upgrade pip
-	$(PIP) install -r requirements.txt
+DOCKER = docker
+TESTDB = solver-rewards-test-db
 
+# Write a marker .install file to indicate that the dependencies have been
+# installed.
+INST := $(VENV)/.install
+$(INST): requirements.txt
+	$(PYTHON) -m venv $(VENV)
+	$(ACTIVATE); pip install --upgrade pip
+	$(ACTIVATE); pip install -r requirements.txt
+	touch $@
 
-install:
-	make $(VENV)/bin/activate
+.PHONY: install
+install: $(INST)
 
+.PHONY: clean
 clean:
-	rm -rf __pycache__
+	rm -rf __pycache__ venv
 
-fmt:
-	black ./
+.PHONY: fmt
+fmt: install
+	$(ACTIVATE); black ./
 
-lint:
-	pylint src/
+.PHONY: lint
+lint: install
+	$(ACTIVATE); pylint src/
 
-types:
-	mypy src/ --strict
+.PHONY: types
+types: install
+	$(ACTIVATE); mypy src/ --strict
 
-check:
-	make install
-	make fmt
-	make lint
-	make types
+.PHONY: check
+check: fmt lint types
 
-test-unit:
-	python -m pytest tests/unit
+.PHONY: test-unit
+test-unit: install
+	$(ACTIVATE); python -m pytest tests/unit
 
-test-e2e:
-	python -m pytest tests/e2e
+.PHONY: test-e2e
+test-e2e: install
+	$(ACTIVATE); python -m pytest tests/e2e
 
+.PHONY: db
 db:
-	# TODO - Should only build if necessary.
```

Comment:
Removed this comment since `docker build` already does this internally (i.e. it will effectively not do anything if you run this multiple times, it reuses existing layers/images that it already built).